### ISSUE TITLE
Add per-bundle token info to hook_token_info()

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#5707]
## Description of the need

While working on https://github.com/backdrop/backdrop-issues/issues/2949 I wanted the ability to add field tokens to only the bundles where the fields were used. But AFAICT there is not currently any way to add token info for a specific bundle. 

I don't remember this being a problem in Drupal, but I couldn't find any similar issues where this has already been solved.

## Proposed solution

In `hook_token_info()` :
* Add a new (optional) 'sub_types' key to the `types` associative array. This key would define another associative array containing sub types of tokens (bundle names, for example). 
* Add a new (optional) 'sub_types' key for each token defined in the associative array of `tokens`. 

These sub_types can later be used to limit the availability of those tokens, perhaps in the token browser UI. 

On the URL Alias Patterns page, we could provide a specific browser for each alias type. The browser for `Post`, for example, would include all the tokens for Post fields, but none of the tokens for Page fields. The browser for the general `Content` alias type would be the same one we have now (with all the possible options - including some that may not exist).

Content returned from hook_token_info() could look something like this:

```php
function node_token_info() {
  $types = array(
    'node' => array(
      'name' => t('Content'),
      'description' => t('Tokens related to individual pieces of content, or "nodes".'),
      'needs-data' => 'node',
      'sub_types' => array('page', 'post', 'book'),
    ),
    ...
  );
...
}
function book_token_info() {
  $info['tokens']['node']['book'] = array(
    'name' => t('Book'),
    'description' => t('The book page associated with the node.'),
    'type' => 'menu-link',
    'sub_types' => array('book'),
  );
  return $info;
}
```

## Alternatives that have been considered

* ?

## Draft of feature description for Press Release (1 paragraph at most)

Backdrop now includes subtypes for tokens.

**Comment by argiepiano** [comment]
This suggestion makes sense to me, but here are some thoughts to start further discussion:

- Token types are not limited to entities (e.g. `site` tokens). For those token types, having a subtype that refers to a bundle is not needed
- There exist tokens for different types of "elements": Field API fields, entity properties, and custom values that are calculated on the fly. The only token for which a bundle differentiation makes sense are Field API field tokens. 

Rather than including the bundle as a subkey within the token definition array, I would suggest an approach similar to Views fields, which in many ways are similar to tokens (the may exist for non-entities, and also can define "views fields" that can be non-Field-API fields). In the case of Views, the bundles are included only as informational in the UI (see below), and are calculated by the UI controller (not coded as a subkey in the field definition).

Implementing it this way may be more backward compatible than adding a new key. You can also limit this bundle checking only for Field API field tokens (which is basically the only tokens where this is needed).

<img width="730" alt="Screen Shot 2022-08-07 at 8 45 39 AM" src="https://user-images.githubusercontent.com/9938978/183296654-ef7fd6eb-21e6-4d39-97b8-a6c53d437ba3.png">

**Comment by argiepiano** [comment]
Unless, of course, I'm misunderstanding the original motivation for this post. I'm unclear what you mean by 

> add field tokens to only the bundles where the fields were used

**Comment by docwilmot** [comment]
I find tokens a bit confusing myself, but the way I read this, I'd say this is "works as designed". Token shortcodes aren't programmatically entered anywhere, somebody's go to type them into a field to use them. In a custom form or a node body, I can type `[node:whatever_nonsense_i_want]`, its my job to make sure it exists. So I'm also unclear what you mean by "add field tokens to only the bundles where the fields were used"

**Comment by jenlampton** [comment]
> Token types are not limited to entities (e.g. site tokens). For those token types, having a subtype that refers to a bundle is not needed

Right, that's why I called it a sub_type (and not a bundle). This new parameter would need to be optional for b/c anyway.  in core we'd only add bundles for the entity related tokens. 

> There exist tokens for different types of "elements": Field API fields, entity properties, and custom values that are calculated on the fly. The only token for which a bundle differentiation makes sense are Field API field tokens.

Not necessarily, in my code example I used book, because a book token would be specific to only node types that were book-enabled. Comment could similarly be limited by node types that have comments. I'm not sure how strict any of this needs to be enforced yet, but I'd like to have the option to limit some tokens. 

> Rather than including the bundle as a subkey within the token definition array, I would suggest an approach similar to Views fields

Interesting...  can you provide an example of what this might look like? What kind of API addition can we make to facilitate this kind of change?

> I'd say this is "works as designed". 

You're right, it does work as it was designed. But I believe we can improve it :) 

>  In a custom form or a node body, I can type [node:whatever_nonsense_i_want], its my job to make sure it exists.

I don't like that the token browser recommends tokens that can not exist. It's one thing if you type in something random, but in this case the recommendation provided by the browser is wrong, even when we have the configuration to know better. Backdrop should know that a token for a field cannot exist if that field does not exist in that location.

You can see an example of this on the path alias patterns page as long as you'd have some kind of chained tokens. You can either use the core patch from https://github.com/backdrop/backdrop-issues/issues/2949 or install the entity_tokens module. 

If you have either of those, you'll still see recommended tokens like `[node:field_term_reference:field_example]`  when field_term_reference is limited to Vocabulary A, and Vocabulary A does not have field_example. 

In this case, when generating a list of tokens for the related entity (term), we could pass the bundle (vocabulary) information along. This could prevent the impossible tokens from being listed in the interface.

(@BWPanda did a much better job of explaining this in https://github.com/backdrop/backdrop-issues/issues/2949#issuecomment-680554510)

---

Right now, on path alias patterns page we have a single link to the token browser for each section (one for node, one for user, etc) because we don't have the ability to build a more specific browser. 

Instead, we could provide a specific browser for each alias type. The browser for `Post`, for example, would include all the tokens for Post fields, but none of the tokens for Page fields. The browser for the general `Content` alias type would be the same one we have now (with all the possible options - including some that may not exist).
